# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: xpcom/base/AppShutdown.cpp
# Commit: 679b52ee70c7
# Full Hash: 679b52ee70c726a41b2522efe9381f7b7534ab3a
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2020-02-18 21:33:59
# Description:
#   Bug 1616186 - Don't try to write StartupCache if it doesn't exist r=froydnj
#   
#   The exact circumstances of how this is showing up in the wild aren't
#   clear - there seem to be a couple of ways we can get here. However it
#   all revolves around early shutdowns (i.e., from the select profile popup)
# ==============================================================================

diff -r ab8ec6eab00c -r 679b52ee70c7 xpcom/base/AppShutdown.cpp
--- a/xpcom/base/AppShutdown.cpp	Tue Feb 18 15:56:33 2020 +0000
+++ b/xpcom/base/AppShutdown.cpp	Tue Feb 18 17:49:47 2020 +0000
@@ -127,7 +127,11 @@
       StaticPrefs::toolkit_shutdown_lateWriteChecksStage();
   sLateWriteChecksPhase = GetShutdownPhaseFromPrefValue(lateWriteChecksPref);
 
-  scache::StartupCache::GetSingleton()->MaybeInitShutdownWrite();
+  // Very early shutdowns can happen before the startup cache is even
+  // initialized; don't bother initializing it during shutdown.
+  if (auto* cache = scache::StartupCache::GetSingletonNoInit()) {
+    cache->MaybeInitShutdownWrite();
+  }
 }
 
 void AppShutdown::MaybeFastShutdown(ShutdownPhase aPhase) {
